# **Comparing the performance of supervised machine learning algorithms**

**Homework 1**

**Course: CSIT 598 Machine Learning**

**Student: Shibbir Ahmed Arif**

**Objective**

In this assignment, we will implement and evaluate several classification algorithms on the MNIST dataset https://yann.lecun.com/exdb/mnist/. The goal is to compare the performance of different machine learning algorithms and analyze their strengths and weaknesses.

**Dataset**

We will work with the MNIST dataset (handwritten digits), which has been widely used in research for benchmarking. We may use preprocessed versions of the dataset available through libraries such as TensorFlow or scikit-learn for MNIST.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import tensorflow as tf
from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

**Data Load**

In [2]:
mnist = tf.keras.datasets.mnist

In [3]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


**Normalize the images**

In [4]:
X_train = X_train.astype('float32') / 255.0
X_test = X_test.astype('float32') / 255.0

**Flatten the images for the classifiers**

In [5]:
X_train = X_train.reshape(-1, 28 * 28)
X_test = X_test.reshape(-1, 28 * 28)

**1. Decision Tree Classifier**

In [6]:
dt_classifier = DecisionTreeClassifier()
dt_classifier.fit(X_train, y_train)

dt_predictions = dt_classifier.predict(X_test)

dt_accuracy = accuracy_score(y_test, dt_predictions)
dt_precision = precision_score(y_test, dt_predictions, average='macro')
dt_recall = recall_score(y_test, dt_predictions, average='macro')
dt_f1score = f1_score(y_test, dt_predictions, average='macro')

print(f'Decision Tree Accuracy: {dt_accuracy:.2f}')
print(f'Decision Tree Precision: {dt_precision:.2f}')
print(f'Decision Tree Recall: {dt_recall:.2f}')
print(f'Decision Tree F1 Score: {dt_f1score:.2f}')

Decision Tree Accuracy: 0.88
Decision Tree Precision: 0.88
Decision Tree Recall: 0.88
Decision Tree F1 Score: 0.88


**2. Naive Bayes Classifier**

**Gaussian Naive Bayes**

In [7]:
gnb_classifier = GaussianNB()
gnb_classifier.fit(X_train, y_train)

gnb_predictions = gnb_classifier.predict(X_test)

gnb_accuracy = accuracy_score(y_test, gnb_predictions)
gnb_precision = precision_score(y_test, gnb_predictions, average='macro')
gnb_recall = recall_score(y_test, gnb_predictions, average='macro')
gnb_f1score = f1_score(y_test, gnb_predictions, average='macro')

print(f'Gaussian Naive Bayes Accuracy: {gnb_accuracy:.2f}')
print(f'Gaussian Naive Bayes Precision: {gnb_precision:.2f}')
print(f'Gaussian Naive Bayes Recall: {gnb_recall:.2f}')
print(f'Gaussian Naive Bayes F1 Score: {gnb_f1score:.2f}')

Gaussian Naive Bayes Accuracy: 0.56
Gaussian Naive Bayes Precision: 0.69
Gaussian Naive Bayes Recall: 0.55
Gaussian Naive Bayes F1 Score: 0.51


**Multinomial Naive Bayes**

In [8]:
mnb_classifier = MultinomialNB()
mnb_classifier.fit(X_train, y_train)

mnb_predictions = mnb_classifier.predict(X_test)

mnb_accuracy = accuracy_score(y_test, mnb_predictions)
mnb_precision = precision_score(y_test, mnb_predictions, average='macro')
mnb_recall = recall_score(y_test, mnb_predictions, average='macro')
mnb_f1score = f1_score(y_test, mnb_predictions, average='macro')

print(f'Multinomial Naive Bayes Accuracy: {mnb_accuracy:.2f}')
print(f'Multinomial Naive Bayes Precision: {mnb_precision:.2f}')
print(f'Multinomial Naive Bayes Recall: {mnb_recall:.2f}')
print(f'Multinomial Naive Bayes F1 Score: {mnb_f1score:.2f}')

Multinomial Naive Bayes Accuracy: 0.84
Multinomial Naive Bayes Precision: 0.84
Multinomial Naive Bayes Recall: 0.83
Multinomial Naive Bayes F1 Score: 0.83


**Comparison of Gaussian and Multinomial Naive Bayes**

In [9]:
# dictionary for the metrics
data = {
    "Metric": ["Accuracy", "Precision", "Recall", "F1 Score"],
    "Gaussian NB": [gnb_accuracy, gnb_precision, gnb_recall, gnb_f1score],
    "Multinomial NB": [mnb_accuracy, mnb_precision, mnb_recall, mnb_f1score]
}

results_df = pd.DataFrame(data)

results_df.set_index("Metric", inplace=True)

results_df = results_df.map(lambda x: f"{x:.2f}")

print(results_df)

          Gaussian NB Multinomial NB
Metric                              
Accuracy         0.56           0.84
Precision        0.69           0.84
Recall           0.55           0.83
F1 Score         0.51           0.83


**3. Support Vector Machine (SVM)**

**Linear SVM**

In [10]:
svm_linear = SVC(kernel='linear')
svm_linear.fit(X_train, y_train)

svm_linear_predictions = svm_linear.predict(X_test)

svm_linear_accuracy = accuracy_score(y_test, svm_linear_predictions)
svm_linear_precision = precision_score(y_test, svm_linear_predictions, average='macro')
svm_linear_recall = recall_score(y_test, svm_linear_predictions, average='macro')
svm_linear_f1score = f1_score(y_test, svm_linear_predictions, average='macro')

**RBF SVM**

In [11]:
svm_rbf = SVC(kernel='rbf')
svm_rbf.fit(X_train, y_train)

svm_rbf_predictions = svm_rbf.predict(X_test)

svm_rbf_accuracy = accuracy_score(y_test, svm_rbf_predictions)
svm_rbf_precision = precision_score(y_test, svm_rbf_predictions, average='macro')
svm_rbf_recall = recall_score(y_test, svm_rbf_predictions, average='macro')
svm_rbf_f1score = f1_score(y_test, svm_rbf_predictions, average='macro')

**Comparison of SVM Kernels**

In [12]:
# dictionary for the comparison
svm_data = {
    "Metric": ["Accuracy", "Precision", "Recall", "F1 Score"],
    "Linear-SVM": [svm_linear_accuracy, svm_linear_precision, svm_linear_recall, svm_linear_f1score],
    "RBF-SVM": [svm_rbf_accuracy, svm_rbf_precision, svm_rbf_recall, svm_rbf_f1score]
}

svm_results_df = pd.DataFrame(svm_data)

svm_results_df.set_index("Metric", inplace=True)

svm_results_df = svm_results_df.map(lambda x: f"{x:.2f}")

print(svm_results_df)

          Linear-SVM RBF-SVM
Metric                      
Accuracy        0.94    0.98
Precision       0.94    0.98
Recall          0.94    0.98
F1 Score        0.94    0.98


**4. k-Nearest Neighbors (k-NN)**

In [13]:
k_values = list(range(1, 10))
accuracy_scores = []
precision_scores = []
recall_scores = []
f1_scores = []

for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)

    knn.fit(X_train, y_train)

    knn_predictions = knn.predict(X_test)

    accuracy = accuracy_score(y_test, knn_predictions)
    precision = precision_score(y_test, knn_predictions, average='macro')
    recall = recall_score(y_test, knn_predictions, average='macro')
    f1 = f1_score(y_test, knn_predictions, average='macro')

    accuracy_scores.append(accuracy)
    precision_scores.append(precision)
    recall_scores.append(recall)
    f1_scores.append(f1)

In [14]:
knn_data = {
    "Metric": ["Accuracy", "Precision", "Recall", "F1 Score"]
}

for i, k in enumerate(k_values):
    knn_data[f"k={k}"] = [
        accuracy_scores[i],
        precision_scores[i],
        recall_scores[i],
        f1_scores[i]
    ]


knn_results_df = pd.DataFrame(knn_data)

knn_results_df.set_index("Metric", inplace=True)

knn_results_df = knn_results_df.map(lambda x: f"{x:.2f}")

print(knn_results_df)

            k=1   k=2   k=3   k=4   k=5   k=6   k=7   k=8   k=9
Metric                                                         
Accuracy   0.97  0.96  0.97  0.97  0.97  0.97  0.97  0.97  0.97
Precision  0.97  0.96  0.97  0.97  0.97  0.97  0.97  0.97  0.97
Recall     0.97  0.96  0.97  0.97  0.97  0.97  0.97  0.97  0.97
F1 Score   0.97  0.96  0.97  0.97  0.97  0.97  0.97  0.97  0.97


**5. Ensemble Methods**

**Bagging with Decision Trees**

In [15]:
bagging_classifier = BaggingClassifier(estimator=DecisionTreeClassifier(), n_estimators=100, random_state=42)
bagging_classifier.fit(X_train, y_train)

bagging_predictions = bagging_classifier.predict(X_test)

bagging_accuracy = accuracy_score(y_test, bagging_predictions)
bagging_precision = precision_score(y_test, bagging_predictions, average='macro')
bagging_recall = recall_score(y_test, bagging_predictions, average='macro')
bagging_f1score = f1_score(y_test, bagging_predictions, average='macro')

print(f'Bagging Accuracy: {bagging_accuracy:.2f}')
print(f'Bagging Precision: {bagging_precision:.2f}')
print(f'Bagging Recall: {bagging_recall:.2f}')
print(f'Bagging F1 Score: {bagging_f1score:.2f}')

Bagging Accuracy: 0.96
Bagging Precision: 0.96
Bagging Recall: 0.96
Bagging F1 Score: 0.96


**Random Forest**

In [16]:
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_classifier.fit(X_train, y_train)

rf_predictions = rf_classifier.predict(X_test)

rf_accuracy = accuracy_score(y_test, rf_predictions)
rf_precision = precision_score(y_test, rf_predictions, average='macro')
rf_recall = recall_score(y_test, rf_predictions, average='macro')
rf_f1score = f1_score(y_test, rf_predictions, average='macro')

print(f'Random Forest Accuracy: {rf_accuracy:.2f}')
print(f'Random Forest Precision: {rf_precision:.2f}')
print(f'Random Forest Recall: {rf_recall:.2f}')
print(f'Random Forest F1 Score: {rf_f1score:.2f}')

Random Forest Accuracy: 0.97
Random Forest Precision: 0.97
Random Forest Recall: 0.97
Random Forest F1 Score: 0.97


**AdaBoost with Decision Trees**

In [19]:
adaboost_classifier = AdaBoostClassifier(estimator=DecisionTreeClassifier(), n_estimators=100, random_state=42)
adaboost_classifier.fit(X_train, y_train)

adaboost_predictions = adaboost_classifier.predict(X_test)

adaboost_accuracy = accuracy_score(y_test, adaboost_predictions)
adaboost_precision = precision_score(y_test, adaboost_predictions, average='macro')
adaboost_recall = recall_score(y_test, adaboost_predictions, average='macro')
adaboost_f1score = f1_score(y_test, adaboost_predictions, average='macro')

print(f'AdaBoost Accuracy: {adaboost_accuracy:.2f}')
print(f'AdaBoost Precision: {adaboost_precision:.2f}')
print(f'AdaBoost Recall: {adaboost_recall:.2f}')
print(f'AdaBoost F1 Score: {adaboost_f1score:.2f}')

AdaBoost Accuracy: 0.88
AdaBoost Precision: 0.87
AdaBoost Recall: 0.87
AdaBoost F1 Score: 0.87


In [20]:
data = {
    "Metric": ["Accuracy", "Precision", "Recall", "F1 Score"],
    "Bagging": [bagging_accuracy, bagging_precision, bagging_recall, bagging_f1score],
    "Random Forest": [rf_accuracy, rf_precision, rf_recall, rf_f1score],
    "AdaBoost": [adaboost_accuracy, adaboost_precision, adaboost_recall, adaboost_f1score]
}

results_df = pd.DataFrame(data)

results_df.set_index("Metric", inplace=True)

results_df = results_df.map(lambda x: f"{x:.2f}")

print(results_df)

          Bagging Random Forest AdaBoost
Metric                                  
Accuracy     0.96          0.97     0.88
Precision    0.96          0.97     0.87
Recall       0.96          0.97     0.87
F1 Score     0.96          0.97     0.87


- Random Forest performs best across all metrics (Accuracy, Precision, Recall, F1 Score) at 0.97, making it the top choice.
- Bagging closely follows with 0.96 across metrics, showing strong but slightly lower performance.
- AdaBoost lags with 0.87 in all metrics, indicating it is less effective on this dataset